### PR TITLE
chore(linter): fixes for rule 16

### PR DESCRIPTION
**Pull request by tarekziade**
# What does this PR do?

Adds support for rules 16

### CODE REVIEW

**Comment by tarekziade** [comment]
run-slow: albert, align, altclip, autoformer, bit, blt, bridgetower, chinese_clip, clap, clip, clipseg, clvp, colmodernvbert, colpali, colqwen2, conditional_detr

File: utils/modular_model_converter.py [hunk 1]
```diff
@@ -368,14 +368,23 @@ def is_call_to_super(self, node: cst.BaseStatement, func_name: str):
         super_call_node = m.Call(func=m.Attribute(value=m.Call(func=m.Name("super")), attr=m.Name(func_name)))
         return m.matches(node, m.SimpleStatementLine(body=[m.Return(super_call_node) | m.Expr(super_call_node)]))
 
+    # Methods whose `super().method(...)` calls should be preserved verbatim instead of
+    # being treated as "inline parent's body here" directives. TRF018 requires a literal
+    # chain to the base class for `_init_weights`.
+    LITERAL_SUPER_METHODS: frozenset[str] = frozenset({"_init_weights"})
```
Comment (zucchini-nlp) [inline utils/modular_model_converter.py]: i am not sure about this part, since it means we won't be able to copy `init_weights` from the parent anymore. 

In a modular file we check for `PreTrainedModel._init_weights(module)` or checking that the body exists iirc, what was the complain from mlinter?

File: src/transformers/models/gemma4/modeling_gemma4.py [hunk 2]
```diff
@@ -1471,6 +1471,7 @@ class Gemma4PreTrainedModel(PreTrainedModel):
 
     @torch.no_grad()
     def _init_weights(self, module):
+        super()._init_weights(module)
         super()._init_weights(module)
```
Comment (hmellor) [inline src/transformers/models/gemma4/modeling_gemma4.py]: 🤔

File: utils/rules.toml [hunk 3]
```diff
@@ -249,3 +249,176 @@ diff = '''
...
+    "vjepa2",
+    "voxtral_realtime",
+    "wav2vec2",
+    "wav2vec2_bert",
+    "wav2vec2_conformer",
+    "wavlm",
+    "x_clip",
+    "xlm",
+    "xlstm",
+]
```
Comment (vasqu) [inline utils/rules.toml]: that's quite a few skips :D let's iterate fast then

File: utils/rules.toml
Comment (vasqu) [inline utils/rules.toml:1]: Different issue but mentioning https://github.com/huggingface/transformers-mlinter/issues/5#top just to keep track

**Review by vasqu — APPROVED** [review]
I'm good with the current changes. 

It kind of remains how we want to handle TRF 018: Imo, it would make sense to call this in any case - the performance loss is negligible for much cleaner modular paths in the future; any init already covered in the super can be overwritten anyway. But I'm not sure what @zucchini-nlp meant with the weird init patterns because they probably have been removed (by limiting the model scope) --> is there anything that would interfere? Maybe we could even remove some small overrides if they already do what the super call does

File: src/transformers/models/gemma4/modeling_gemma4.py (hunk 2 above)
Comment (tarekziade) [reply src/transformers/models/gemma4/modeling_gemma4.py]: We need to be extra sure! ;)

File: utils/rules.toml (hunk 3 above)
Comment (tarekziade) [reply utils/rules.toml]: I am happy to push all of the ~190 models in the same PR, let me know

File: utils/modular_model_converter.py (hunk 1 above)
Comment (tarekziade) [reply utils/modular_model_converter.py]: Inlining `super()._init_weights(module)` is the bug TRF018 catches: the generated file ends up with the parent's body copied in but no chain to `PreTrainedModel._init_weights`, silently  breaking init

Keeping the call literal preserves the chain at runtime via MRO and satisfies the linter. If a model genuinely wants to skip the chain, `# trf-ignore: TRF018` is the escape hatch.

File: utils/rules.toml (hunk 3 above)
Comment (vasqu) [reply utils/rules.toml]: I think we could do splits based on this branch and only have them add smaller batches; then we merge them into this branch bit by bit. Wdyt?

Comment (tarekziade) [reply utils/rules.toml]: you mean like separate commits in this branch ?

Comment (vasqu) [reply utils/rules.toml]: I thought of minimal branches that only handle a subset and those should be easy to merge into this branch then. The idea is basically to have only small subsets we have to go through each time. But if that's too much, we can also push everything at once no worries

File: src/transformers/models/gemma4/video_processing_gemma4.py [hunk 4]
```diff
@@ -104,7 +104,6 @@ class Gemma4VideoProcessor(BaseVideoProcessor):
     image_std = [1.0, 1.0, 1.0]
     size = None
     default_to_square = True
-    do_convert_rgb = True
```
Comment (zucchini-nlp) [inline src/transformers/models/gemma4/video_processing_gemma4.py:107]: why is it being removed btw? i see a few more deletes below for do pad, resize

**Review by zucchini-nlp — COMMENTED** [review]
Re init weights thing, i followed the slack thread with Cyril and prob misunderstood that init weights once via super will set the flag on the weight already. Since it can be init correctly even after calling super, I am fine with it

My only q is about deleted `do_xxx` in some processors. Maybe we can split into 3 PRs for each rule, imo we don't need to PR small subsets of models if we are happy with calling `super().init_weights` always

File: src/transformers/models/gemma4/video_processing_gemma4.py (hunk 4 above)
Comment (vasqu) [reply src/transformers/models/gemma4/video_processing_gemma4.py:107]: I think that's `TRF016` kicking in here, it's not referenced. Maybe we need more edge cases?

Comment (zucchini-nlp) [reply src/transformers/models/gemma4/video_processing_gemma4.py:107]: hmm indeed, prob they assumed it happens automatically outside `prerpcessor`, because that is how image processing works. I will pass by processing changes one-by-one to verify it can actually be deleted then

**Comment by tarekziade** [comment]
splited out rules 17 and 18 on their own PR

https://github.com/huggingface/transformers/pull/46260
https://github.com/huggingface/transformers/pull/46259

**Comment by tarekziade** [comment]
run-slow: gemma4, glm46v, glm4v, glmga, kosmos2_5, llama4, mask2former, maskformer, minicpmv4_6, mllama, oneformer, perception_lm, phi4_multimodal, qwen2_vl, qwen3_vl, video_llama_3

File: docs/source/en/modeling_rules.md [hunk 5]
```diff
@@ -228,6 +228,61 @@ When a PreTrainedModel subclass defines _tied_weights_keys as a non-empty collec
 +    tie_word_embeddings: bool = True
 ```
 
+### TRF016
```
Comment (zucchini-nlp) [inline docs/source/en/modeling_rules.md]: i suppose the docs contains unrelated changes

File: src/transformers/models/gemma4/modeling_gemma4.py
Comment (zucchini-nlp) [inline src/transformers/models/gemma4/modeling_gemma4.py:594]: unrelated

File: src/transformers/models/kosmos2_5/modeling_kosmos2_5.py
Comment (zucchini-nlp) [inline src/transformers/models/kosmos2_5/modeling_kosmos2_5.py:639]: also not related

File: src/transformers/models/glmga/video_processing_glmga.py [hunk 6]
```diff
@@ -70,7 +70,6 @@ class GlmgaVideoProcessor(BaseVideoProcessor):
     do_resize = True
     do_rescale = True
     do_normalize = True
-    do_convert_rgb = True
```
Comment (zucchini-nlp) [inline src/transformers/models/glmga/video_processing_glmga.py:73]: seeing all video processors don't use `donvert_rgb`. I suppose the contribs assumed it gets caught up internally

imo we have to keep the flag and make sure it is actually used. We need to squeeze it in here:

```
if do_convert_rgb:
    video = self.convert_to_rgb(video)
```

https://github.com/huggingface/transformers/blob/455709256d731fd597995877a9c5a88c479fe38e/src/transformers/video_processing_utils.py#L321-L324

File: tests/models/kosmos2_5/test_modeling_kosmos2_5.py
Comment (zucchini-nlp) [inline tests/models/kosmos2_5/test_modeling_kosmos2_5.py:495]: unrelated as well

File: src/transformers/models/llama4/image_processing_llama4.py [hunk 7]
```diff
@@ -298,7 +298,6 @@ class Llama4ImageProcessor(TorchvisionBackend):
     image_mean = [0.5, 0.5, 0.5]
     image_std = [0.5, 0.5, 0.5]
     size = {"height": 336, "width": 336}
-    do_resize = True
```
Comment (zucchini-nlp) [inline src/transformers/models/llama4/image_processing_llama4.py:301]: can we TRF ignore it? imo it is meant to always resize and we can actually check if `do_resize=False` and raise an error that it cannot be False, otherwise the output pixel values cannot be tensorized

File: src/transformers/models/oneformer/image_processing_oneformer.py [hunk 8]
```diff
@@ -296,7 +296,6 @@ class OneFormerImageProcessor(TorchvisionBackend):
     do_rescale = True
     do_normalize = True
     default_to_square = False
-    do_center_crop = False
```
Comment (zucchini-nlp) [inline src/transformers/models/oneformer/image_processing_oneformer.py:299]: can also delete `crop_size = None`, same for PIL backend

**Review by zucchini-nlp — COMMENTED** [review]
A few comments and lets' revert unrelated commits

File: docs/source/en/modeling_rules.md (hunk 5 above)
Comment (tarekziade) [reply docs/source/en/modeling_rules.md]: changes in this file and rules.toml will be removed from the patch. we just want to make the code base rule-16-compatible so we can update mlinter to the next release in another patch

File: utils/modular_model_converter.py (hunk 1 above)
Comment (tarekziade) [reply utils/modular_model_converter.py]: rule 18 related stuff splitted out another PR

**Review by zucchini-nlp — APPROVED** [review]
Great, thanks for iterating!

**Comment by tarekziade** [comment]
run-slow: gemma4, glm46v, glm4v, glmga, llama4, mask2former, maskformer, minicpmv4_6, mllama, oneformer, perception_lm, phi4_multimodal, qwen2_vl, qwen3_vl, video_llama_3

**Review by vasqu — APPROVED** [review]
Thanks 🤗

**Comment by vasqu** [comment]
I think we are safe to merge? I recall some rules.toml etc but ig we dont need them here?
